File: stml_java/src/main/java/com/stml/serializer/Serializer.java
package com.stml.serializer;

import com.stml.ast.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Serializer: AST → canonical STML text / JSON text.
 */
public final class Serializer {

    private Serializer() {}

    // =========================================================================
    // Escape helpers
    // =========================================================================

    /** Escape special characters for double-quoted strings. */
    public static String escapeString(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 8);
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            switch (ch) {
                case '\\': sb.append("\\\\"); break;
                case '"':  sb.append("\\\""); break;
                case '\n': sb.append("\\n"); break;
                case '\t': sb.append("\\t"); break;
                default:   sb.append(ch); break;
            }
        }
        return sb.toString();
    }

    /** JSON string escape (more thorough than STML escape). */
    private static String jsonEscape(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 8);
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            switch (ch) {
                case '\\': sb.append("\\\\"); break;
                case '"':  sb.append("\\\""); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                default:
                    if (ch < 0x20) {
                        sb.append("\\u00");
                        sb.append("0123456789abcdef".charAt(ch >> 4));
                        sb.append("0123456789abcdef".charAt(ch & 0xf));
                    } else {
                        sb.append(ch);
                    }
                    break;
            }
        }
        return sb.toString();
    }

    /** Format a key (always double-quoted). */
    public static String quoteKey(String key) {
        return "\"" + escapeString(key) + "\"";
    }

    // =========================================================================
    // formatScalar
    // =========================================================================

    /** Format a scalar value as STML representation. */
    public static String formatScalar(AstNode node) {
        if (node.isNull()) {
            return "null";
        }
        if (node.isString()) {
            return "\"" + escapeString(node.asString()) + "\"";
        }
        if (node.isMap() && node.asMap().isEmpty()) {
            return "null";
        }
        if (node.isList() && node.asList().isEmpty()) {
            return "null";
        }
        throw new IllegalArgumentException("formatScalar called on non-scalar node");
    }

    // =========================================================================
    // serializeNode
    // =========================================================================

    /** Convert an AST node to an STML fragment (possibly multi-line). Indentation: 2 spaces per level. */
    public static String serializeNode(AstNode node, int indentLevel) {
        String indent = " ".repeat(indentLevel * 2);

        // 1. Null
        if (node.isNull()) {
            return indent + "null";
        }

        // 2. String (scalar)
        if (node.isString()) {
            return indent + "\"" + escapeString(node.asString()) + "\"";
        }

        // 3. List (sequence)
        if (node.isList()) {
            List<AstNode> lst = node.asList();
            if (lst.isEmpty()) {
                return indent + "- null";
            }

            List<String> lineList = new ArrayList<>();
            for (AstNode item : lst) {
                if (item.isNull() || item.isString()) {
                    lineList.add(indent + "- " + formatScalar(item));
                } else if (item.isList()) {
                    List<AstNode> innerLst = item.asList();
                    if (innerLst.isEmpty()) {
                        lineList.add(indent + "- null");
                    } else if (valuesAreAllScalars(innerLst)) {
                        lineList.add(indent + "- " + formatInlineList(innerLst));
                    } else {
                        lineList.add(indent + "- ");
                        lineList.add(serializeNode(item, indentLevel + 1));
                    }
                } else if (item.isMap()) {
                    List<Map.Entry<String, AstNode>> itemMap = item.asMap();
                    if (itemMap.isEmpty()) {
                        lineList.add(indent + "- null");
                        continue;
                    }
                    if (itemMap.size() > 1) {
                        throw new IllegalArgumentException(
                                "STML list entries do not support multi-key dicts. Use single-key mapping wrapper.");
                    }
                    var entry = itemMap.get(0);
                    String key = entry.getKey();
                    AstNode val = entry.getValue();
                    if (val.isList() || val.isMap()) {
                        lineList.add(indent + "- " + quoteKey(key) + ":");
                        lineList.add(serializeNode(val, indentLevel + 1));
                    } else {
                        lineList.add(indent + "- " + quoteKey(key) + ": " + formatScalar(val));
                    }
                }
            }
            return String.join("\n", lineList);
        }

        // 4. Map (mapping)
        if (node.isMap()) {
            List<Map.Entry<String, AstNode>> map = node.asMap();
            if (map.isEmpty()) {
                return indent + "null";
            }

            List<String> lineList = new ArrayList<>();
            for (var entry : map) {
                String key = entry.getKey();
                AstNode val = entry.getValue();
                String kStr = quoteKey(key);
                if (val.isList() || val.isMap()) {
                    String sub = serializeNode(val, indentLevel + 1);
                    if (sub.isEmpty()) {
                        lineList.add(indent + kStr + ": null");
                    } else {
                        lineList.add(indent + kStr + ":");
                        lineList.add(sub);
                    }
                } else {
                    lineList.add(indent + kStr + ": " + formatScalar(val));
                }
            }
            return String.join("\n", lineList);
        }

        throw new IllegalArgumentException("Unsupported node type in serializeNode");
    }

    private static boolean valuesAreAllScalars(List<AstNode> lst) {
        for (AstNode v : lst) {
            if (v.isList() || v.isMap()) return false;
        }
        return true;
    }

    private static String formatInlineList(List<AstNode> lst) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < lst.size(); i++) {
            if (i > 0) sb.append(", ");
            sb.append(formatScalar(lst.get(i)));
        }
        sb.append("]");
        return sb.toString();
    }

    // =========================================================================
    // dumps — top-level serialization
    // =========================================================================

    /** Convert docs-wrapped AST to canonical STML text. */
    public static String dumps(AstNode node) {
        if (node.isMap()) {
            AstNode docsNode = AstNode.mapFind(node, "docs");
            if (docsNode != null && docsNode.isList()) {
                List<AstNode> docs = docsNode.asList();
                List<String> parts = new ArrayList<>();
                for (AstNode doc : docs) {
                    parts.add(serializeNode(doc, 0));
                }
                StringBuilder result = new StringBuilder();
                for (int i = 0; i < parts.size(); i++) {
                    result.append(parts.get(i));
                    if (i + 1 < parts.size())
                        result.append("\n---");
                    result.append("\n");
                }
                return result.toString();
            }
        }

        // Unwrapped single document
        return serializeNode(node, 0) + "\n";
    }

    // =========================================================================
    // toJson — AST → canonical JSON text
    // =========================================================================

    /** Convert an AST node to JSON text. */
    public static String toJson(AstNode node) {
        StringBuilder sb = new StringBuilder();
        toJsonImpl(node, sb, 0);
        sb.append("\n");
        return sb.toString();
    }

    private static void toJsonImpl(AstNode node, StringBuilder sb, int indentLevel) {
        if (node.isNull()) {
            sb.append("null");
        } else if (node.isString()) {
            sb.append('"').append(jsonEscape(node.asString())).append('"');
        } else if (node.isList()) {
            List<AstNode> lst = node.asList();
            if (lst.isEmpty()) {
                sb.append("[]");
            } else {
                sb.append("[\n");
                for (int i = 0; i < lst.size(); i++) {
                    writeIndent(sb, indentLevel + 1);
                    toJsonImpl(lst.get(i), sb, indentLevel + 1);
                    if (i + 1 < lst.size()) sb.append(",");
                    sb.append("\n");
                }
                writeIndent(sb, indentLevel);
                sb.append("]");
            }
        } else if (node.isMap()) {
            List<Map.Entry<String, AstNode>> map = node.asMap();
            if (map.isEmpty()) {
                sb.append("{}");
            } else {
                sb.append("{\n");
                int i = 0;
                for (var entry : map) {
                    writeIndent(sb, indentLevel + 1);
                    sb.append('"').append(jsonEscape(entry.getKey())).append("\": ");
                    toJsonImpl(entry.getValue(), sb, indentLevel + 1);
                    if (++i < map.size()) sb.append(",");
                    sb.append("\n");
                }
                writeIndent(sb, indentLevel);
                sb.append("}");
            }
        }
    }

    private static void writeIndent(StringBuilder sb, int level) {
        for (int i = 0; i < level * 2; i++)
            sb.append(' ');
    }
}

File: IdeaValidationProject/serializer.py
"""STML Serializer: AST → canonical STML text.

Directly migrated from JSON2STML.py to guarantee character-level
compatibility with `output.stml`.  All output rules follow §15.

C++ compatible: no generators, explicit types, list accumulation.
"""

import json
from typing import Any, Dict, List


def escape_string(s: str) -> str:
    """Escape double-quote, backslash, newline, tab."""
    s = s.replace('\\', '\\\\')
    s = s.replace('"', '\\"')
    s = s.replace('\n', '\\n')
    s = s.replace('\t', '\\t')
    return s


def quote_key(key: str) -> str:
    """Keys are always double-quoted."""
    return '"' + escape_string(key) + '"'


def format_scalar(value: Any) -> str:
    """Scalar value → STML representation.

    null → unquoted 'null'
    bool / int / float → double-quoted string
    str → double-quoted with escapes
    """
    if value is None:
        return 'null'
    if isinstance(value, bool):
        return '"' + ('true' if value else 'false') + '"'
    if isinstance(value, (int, float)):
        return '"' + str(value) + '"'
    if isinstance(value, str):
        return '"' + escape_string(value) + '"'
    raise TypeError(f"Unsupported type: {type(value)}")


def serialize_node(node: Any, indent_level: int = 0) -> str:
    """Convert an AST node to a (possibly multi-line) STML fragment.
    Indentation: 2 spaces per level.
    """
    indent = '  ' * indent_level
    next_indent = '  ' * (indent_level + 1)

    # 1. Null / scalar
    if node is None or isinstance(node, (bool, int, float, str)):
        return indent + format_scalar(node)

    def _values_are_all_scalars(lst: List[Any]) -> bool:
        """Check if every element in *lst* is a simple scalar (str, None, bool, int, float)."""
        for v in lst:
            if isinstance(v, (dict, list)):
                return False
        return True

    def _format_inline_list(lst: List[Any]) -> str:
        """Format a list of scalars as an inline list: ["v1", "v2", "v3"]."""
        parts: List[str] = []
        for v in lst:
            parts.append(format_scalar(v))
        return '[' + ', '.join(parts) + ']'

    # 2. List (sequence)
    if isinstance(node, list):
        if not node:
            return indent + '- null'
        lines: List[str] = []
        for item in node:
            if isinstance(item, (type(None), bool, int, float, str)):
                # Scalar entry: - "value"
                lines.append(indent + '- ' + format_scalar(item))
            elif isinstance(item, list):
                # Nested list item
                if not item:
                    lines.append(indent + '- null')
                elif _values_are_all_scalars(item):
                    # Inline list: - [v1, v2, v3]
                    lines.append(indent + '- ' + _format_inline_list(item))
                else:
                    # Complex nested list: sub-block
                    lines.append(indent + '- ')
                    lines.append(serialize_node(item, indent_level + 1))
            elif isinstance(item, dict):
                if not item:
                    lines.append(indent + '- null')
                    continue
                if len(item) > 1:
                    raise ValueError(
                        "SYML list entries do not support multi-key dicts. "
                        "Use single-key mapping wrapper: " + str(item)
                    )
                key, val = next(iter(item.items()))
                if isinstance(val, (dict, list)):
                    # Compound value: - "key":
                    lines.append(indent + '- ' + quote_key(key) + ':')
                    sub = serialize_node(val, indent_level + 1)
                    lines.append(sub)
                else:
                    # Scalar value: - "key": "value"
                    lines.append(
                        indent + '- ' + quote_key(key) + ': ' + format_scalar(val)
                    )
            else:
                raise TypeError(f"List contains unsupported type: {type(item)}")
        return '\n'.join(lines)

    # 3. Dict (mapping)
    if isinstance(node, dict):
        if not node:
            return indent + 'null'
        lines: List[str] = []
        for key, val in node.items():
            k_str = quote_key(key)
            if isinstance(val, (dict, list)):
                # Compound value: newline + indent
                sub = serialize_node(val, indent_level + 1)
                if sub.strip() == '':
                    lines.append(indent + k_str + ': null')
                else:
                    lines.append(indent + k_str + ':')
                    lines.append(sub)
            else:
                # Scalar value: same line
                lines.append(indent + k_str + ': ' + format_scalar(val))
        return '\n'.join(lines)

    raise TypeError(f"Unsupported node type: {type(node)}")


def dumps(obj: Dict[str, Any]) -> str:
    """Convert a docs-wrapped AST to canonical STML text.

    Format: {"docs": {"doc1": ..., "doc2": ...}}

    Multiple documents are separated by '---' on its own line.
    """
    if isinstance(obj, dict) and 'docs' in obj and isinstance(obj['docs'], dict):
        docs = obj['docs']
        # Sort keys numerically (doc1, doc2, ...)
        sorted_keys = sorted(docs.keys(), key=lambda x: int(x.replace('doc', '')))
        parts: List[str] = []
        for k in sorted_keys:
            parts.append(serialize_node(docs[k], indent_level=0))
        return '\n---\n'.join(parts) + '\n'
    else:
        # Unwrapped single document
        return serialize_node(obj, indent_level=0) + '\n'


def json_to_stml(json_str: str) -> str:
    """JSON string → STML text."""
    obj = json.loads(json_str)
    return dumps(obj)
